mlflow/store/db_migrations/versions/0584bdc529eb_add_cascading_deletion_to_datasets_from_.py

```diff
@@ -0,0 +1,66 @@
+"""add cascading deletion to datasets from experiments
+
+Revision ID: 0584bdc529eb
+Revises: f5a4f2784254
+Create Date: 2024-11-11 15:27:53.189685
+
+"""
+from alembic import op
+import sqlalchemy as sa
+
+from mlflow.store.tracking.dbmodels.models import SqlDataset, SqlExperiment
+
+
+# revision identifiers, used by Alembic.
+revision = '0584bdc529eb'
+down_revision = 'f5a4f2784254'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    dialect_name = op.get_context().dialect.name
+    if dialect_name == "sqlite":
+        # Only way to drop unnamed fk constraint in sqllite
+        # See https://alembic.sqlalchemy.org/en/latest/batch.html#dropping-unnamed-or-named-foreign-key-constraints
+        with op.batch_alter_table(
+            SqlDataset.__tablename__,
+            schema=None,
+            naming_convention={
+                "fk": "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
+            },
+        ) as batch_op:
+            batch_op.drop_constraint(f"fk_{SqlDataset.__tablename__}_experiment_id_{SqlExperiment.__tablename__}", type_="foreignkey")
+            # Need to explicitly name the fk constraint with batch alter table
+            batch_op.create_foreign_key(
+                f"fk_{SqlDataset.__tablename__}_experiment_id_{SqlExperiment.__tablename__}",
+                SqlExperiment.__tablename__,
+                ["experiment_id"],
+                ["experiment_id"],
+                ondelete="CASCADE",
+            )
+    else:
+        if dialect_name == "postgresql":
+            fk_constraint_name = f"{SqlDataset.__tablename__}_experiment_id_fkey"
+        elif dialect_name == "mysql":
+            fk_constraint_name = f"{SqlDataset.__tablename__}_ibfk_1"
+        elif dialect_name == "mssql":
+            # mssql fk constraint name at f5a4f2784254
+            fk_constraint_name = f"FK__{SqlDataset.__tablename__}__experi__6477ECF3"
+
+        # don't use batch alter table here so `create_foreign_key()` can be
+        # called with `None` as the `constraint_name` and have it set
+        # automatically based on the conventions of the sql flavor
+        op.drop_constraint(fk_constraint_name, SqlDataset.__tablename__, type_="foreignkey")
+        op.create_foreign_key(
+            None,
```

Comment:
Is it possible to name this constraint?

Reply:
i think it's possible but i'm not sure of the implications. maybe we can just standardize on `fk_datasets_experiment_id_experiments"

Reply:
to be safe i feel like it's probably best to do less changes, but i don't have a lot of knowledge about DB stuff. if it's not a problem for the names to change then i'll update the PR!

Reply:
Makes sense!

Reply:
This reminds me of https://github.com/mlflow/mlflow/pull/11852#discussion_r1583428781:

> Can we name the foreign key please? Otherwise MSSQL will generate a generated unique uuid based name that forces additional processing steps for fkey migrations or modifications

Reply:
can we try naming the constraint? `tests/db/check_migration.py` should pass if renaming doesn't work or is not allowed.

Reply:
it looks like it works okay, how do i run that script in a docker container?

Reply:
or can we just wait for CI

Reply:
hmm it's quite weird, running `./tests/db/compose.sh run migration-postgresql` i get the following error during the pre-migration check:

```
mlflow.exceptions.MlflowException: Detected out-of-date database schema (found version 0584bdc529eb, but expected f5a4f2784254).
```

it seems the database is already migrated (`0584bdc529eb` is the revision in this PR), maybe some caching thing on my end? i'm not sure if that's the right command to run

Reply:
Interesting. Let me try.

Reply:
ah awesome, thanks! all looks good then, i'll merge and cherry pick